samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java

```diff
@@ -253,9 +255,10 @@ private static void calculateIntStreamPartitions(JobGraph jobGraph, Config confi
     if (partitions < 0) {
       // use the following simple algo to figure out the partitions
       // partition = MAX(MAX(Input topic partitions), MAX(Output topic partitions))
+      // partition will be further bounded by MAX_INFERRED_PARTITIONS. This is important when running in hadoop.
       int maxInPartitions = maxPartition(jobGraph.getSources());
       int maxOutPartitions = maxPartition(jobGraph.getSinks());
-      partitions = Math.max(maxInPartitions, maxOutPartitions);
+      partitions = Math.min(Math.max(maxInPartitions, maxOutPartitions), MAX_INFERRED_PARTITIONS);
```

Comment:
There should at least be a WARN message if MAX_INFERRED_PARTITIONS, because artificially choosing a max can affect correctness(joins) and stability (too much throughput per partition).

Reply:
+1